addons/mrp/models/mrp_bom.py

```diff
@@ -249,10 +252,16 @@ def check_cycle(v, visited, recStack, graph):
         lines_done = []
         V |= set([product.product_tmpl_id.id])
 
-        bom_lines = [(bom_line, product, quantity, False) for bom_line in self.bom_line_ids]
+        bom_lines = []
+        products_to_search = self.env['product.product']
         for bom_line in self.bom_line_ids:
-            V |= set([bom_line.product_id.product_tmpl_id.id])
-            graph[product.product_tmpl_id.id].append(bom_line.product_id.product_tmpl_id.id)
+            product_id = bom_line.product_id
+            V |= set([product_id.product_tmpl_id.id])
+            graph[product.product_tmpl_id.id].append(product_id.product_tmpl_id.id)
+            bom_lines.append((bom_line, product, quantity, False))
+            products_to_search |= product_id
```

Comment:
I am concerned about the union of the recordset due to his complexity (N²) - (I can be issue if there is more than 5K lines, it is very unlikely, but we are always surprised by the client data quantity). Maybe a OrderedSet of the ids is maybe safer?

Reply:
An id set is better but do we really need an OrderedSet here ? I know that recordsets keep order but here the purpose of products_to_search is to save encountered product_ids for _get_product2bom. Since this function returns an unordered dict wouldn't a simple `set` be enough ?

Reply:
Yes you are perfectly right :+1: , we don't need to conserve the order.